app/src/main/java/com/cmput301f20t21/bookfriends/ui/request/RequestActivity.java

```diff
@@ -0,0 +1,177 @@
+        });
+    }
+
+    /**
+     * display requesters of the book by getting them from FireStore based on bookId
+     * @param bookId
+     */
+    public void displayRequest(String bookId) {
+        requestViewModel.getRequesters(bookId).observe(this, requesters -> {
+            if (requesters != null) {
+                requestDataList.addAll(requesters);
+                requestAdapter.notifyDataSetChanged();
+            }
+        });
+    }
+
+    /**
+     * setup the back button on the title
+     * @param item
+     * @return
+     */
+    @Override
+    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
+        if (item.getItemId() == android.R.id.home) {
+            finish();
+            return true;
+        }
+        return super.onOptionsItemSelected(item);
+    }
+
+    /**
+     * Function to set view by ID, set adapter and build recycler view
+     */
+    public void buildRecyclerView() {
+        recyclerView = findViewById(R.id.request_recycler_view);
+        recyclerView.setHasFixedSize(true);
+        layoutManager = new LinearLayoutManager(this);
+        recyclerView.setLayoutManager(layoutManager);
+        requestAdapter = new RequestAdapter(requestDataList);
+        recyclerView.setAdapter(requestAdapter);
+
+        requestAdapter.setOnItemClickLisener(new RequestAdapter.OnItemClickListener() {
+            @Override
+            public void onRejectClick(int position) {
+                removeItem(position);
+            }
+
+            @Override
+            public void onAcceptClick(int position) {
+                openDialog(position);
+            }
+        });
+    }
+
+    /**
+     * function to remove an item when we click on Reject button
+     * @param position that needs removing
+     */
+    public void removeItem(int position) {
+        Request request = requestDataList.get(position);
+        requestViewModel.removeRequest(request.getId());
+        requestDataList.remove(position);
+        requestAdapter.notifyItemRemoved(position);
+    }
+
+    /**
+     * When user click on the accept button
+     * popup a dialog to prompt user about their action:
+     * accept one item and remove all other items
+     * @param position that we accept the item
+     */
+    public void openDialog(int position) {
+        String acceptedId = requestDataList.get(position).getId();
+        List<String> idsToRemove = new ArrayList<>();
+        for (int i = 0; i < requestDataList.size();i++) {
+            idsToRemove.add(requestDataList.get(i).getId());
+        }
+        idsToRemove.remove(acceptedId);
+        ConfirmDialog confirmDialog = new ConfirmDialog(acceptedId, idsToRemove);
+        confirmDialog.show(getSupportFragmentManager(), "Confirm Dialog");
+    }
+
+    /**
+     * function is called whenever the user confirms to accept a request
+     * remove all other requests
+     */
+    @Override
+    public void setConfirm(String id, List<String> idsToRemove) {
```

Comment:
the idsToRemove didn't change at all, passed back from confirm dialog, exactly the same. The problem here is that 
1. we don't need to pass it at. 
2. we shouldn't create this list inside activity. Instead, idsToRemove is a matter that's the view model's concern. It judges from the request id and the current list of requests, then remove all that needs to be removed on confirm. We shouldn't call `removeAllRequest`, it's not self-contained enought for the vm. instead, a new method `acceptRequestAndRejectOthers(String requestId)` might be needed, and all logic around here should go in there, including the idsToRemove creation logic.

Reply:
if the `removeAllRequest` is needed elsewhere and need to be kept. we should rename it to `removeRequestsByIds` or a more descriptive name. It's not removing all requests